django/template/engine.py

```diff
@@ -52,6 +52,26 @@ def __init__(self, dirs=None, app_dirs=False, context_processors=None,
         self.builtins = self.default_builtins + builtins
         self.template_builtins = self.get_template_builtins(self.builtins)
 
+    def __repr__(self):
+        return (
+            '<%s:%s app_dirs=%s%s debug=%s loaders=%s string_if_invalid=%s '
+            'file_charset=%s%s%s autoescape=%s>'
+        ) % (
+            self.__class__.__qualname__,
+            '' if not self.dirs else ' dirs=%s' % repr(self.dirs),
+            self.app_dirs,
+            ''
+            if not self.context_processors
+            else ' context_processors=%s' % repr(self.context_processors),
+            self.debug,
+            repr(self.loaders),
+            repr(self.string_if_invalid),
+            repr(self.file_charset),
+            '' if not self.libraries else ' libraries=%s' % repr(self.libraries),
+            '' if not self.builtins else ' builtins=%s' % repr(self.builtins),
```

Comment:
I don't think `builtins` can be empty. the `default_builtins` should always be present?

Reply:
They can be empty for subclasses, I think we can leave it that way.

Reply:
okay, then that branch needs a test too i guess?

Reply:
It's only `__repr__()`, I don't think it's worth to test all cases.